library/src/blas2/rocblas_tbmv.hpp

```diff
@@ -0,0 +1,371 @@
+/* ************************************************************************
+ * Copyright 2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#ifndef __ROCBLAS_TBMV_HPP__
+#define __ROCBLAS_TBMV_HPP__
+#include "handle.h"
+#include "rocblas.h"
+#include "utility.h"
+
+// Makes a copy of xa into ya.
+template <typename T, typename U, typename V>
+__global__ void tbmv_copy_helper(rocblas_int    n,
+                                 const U        xa,
+                                 ptrdiff_t      shiftx,
+                                 rocblas_int    incx,
+                                 rocblas_stride stridex,
+                                 V              ya,
+                                 ptrdiff_t      shifty,
+                                 rocblas_int    incy,
+                                 rocblas_stride stridey)
+{
+    ptrdiff_t tid = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    // bound
+    if(tid < n)
+    {
+        const T* x = load_ptr_batch(xa, hipBlockIdx_y, shiftx, stridex);
+        T*       y = load_ptr_batch(ya, hipBlockIdx_y, shifty, stridey);
+
+        y[tid * incy] = x[tid * incx];
+    }
+}
+
+/**
+  *  Helper for the non-transpose case. Iterates through each diagonal
+  *  and creates partial sums for each ty.
+  */
+template <rocblas_int DIM_Y, typename T>
+__device__ T tbmvn_kernel_helper(rocblas_int ty,
+                                 rocblas_int ind,
+                                 bool        upper,
+                                 bool        diag,
+                                 rocblas_int m,
+                                 rocblas_int k,
+                                 const T*    A,
+                                 rocblas_int lda,
+                                 const T*    x_copy,
+                                 rocblas_int incx)
+{
+    T           res_A = 0.0;
+    rocblas_int col   = ty; // ty defines the column of banded & regular matrix
+
+    // Since the column is consistent, we can iterate up the diagonal
+    for(col = ty; col < m; col += DIM_Y)
+    {
+        // We have to convert ind to banded matrix row
+        rocblas_int row = upper ? ind + (k - col) : ind - col;
+
+        if(ind < m)
+        {
+            // Regular case, simply multiply
+            if(row < k && row > 0)
+            {
+                res_A += (A[row + col * lda] * x_copy[col]);
+            }
+            else if(row == 0)
+            {
+                // If main diagonal && diag, don't reference matrix, assume 1.
+                if(diag && !upper)
```

Comment:
minor tweak, not sure if it worth it 
if (diag && (!upper || (k==0 && upper)))

Reply:
Yeah that's a bit cleaner. I'm guessing a lot of this logic can be improved in the future. Changed.